hearthstone/hero.py

```diff
@@ -34,25 +40,39 @@ def tavern_upgrade_costs(self) -> Tuple[int, int, int, int, int, int]:
     def handle_event(self, event: CardEvent, context: Union[BuyPhaseContext, CombatPhaseContext]):
         pass
 
-    def hero_power(self, context: BuyPhaseContext):
-        assert self.hero_power_valid(context)
+    def hero_power(self, context: BuyPhaseContext, board_index: Optional['BoardIndex'] = None,
+                   store_index: Optional['StoreIndex'] = None):
+        assert self.hero_power_valid(context, board_index, store_index)
         context.owner.coins -= self.power_cost
         self.hero_power_used = True
-        self.hero_power_impl(context)
+        self.hero_power_impl(context, board_index, store_index)
 
-    def hero_power_impl(self, context: BuyPhaseContext):
+    def hero_power_impl(self, context: 'BuyPhaseContext', board_index: Optional['BoardIndex'] = None,
+                        store_index: Optional['StoreIndex'] = None):
         pass
 
-    def hero_power_valid(self, context: BuyPhaseContext):
+    def hero_power_valid(self, context: 'BuyPhaseContext', board_index: Optional['BoardIndex'] = None,
+                         store_index: Optional['StoreIndex'] = None):
         if context.owner.coins < self.power_cost:
             return False
         if self.hero_power_used:
             return False
-        if not self.hero_power_valid_impl(context):
+        if not self.can_use_power:
+            return False
+        if board_index is None and store_index is None and self.power_target_location is not None:
+            return False
+        if board_index is not None:
+            if self.power_target_location is not CardLocation.BOARD or not context.owner.valid_board_index(board_index):
```

Comment:
I think you want to use equality here instead of `is`. `is` is generally only necessary when you care about object identities, or are comparing to `None`.